src/net_processing.cpp

```diff
@@ -1809,6 +1818,19 @@ std::set<uint256>& PeerManagerImpl::GetChainTipSetForPeerNetwork(const CNode& no
     return m_chain_tip_sets.emplace(GetUniqueNetworkID(m_connman, node), std::set<uint256>()).first->second;
 }
 
+const CBlockIndex* PeerManagerImpl::LookupBlockIndexForPeer(const CNode& node, const uint256& hash)
+{
+    const CBlockIndex* index{m_chainman.m_blockman.LookupBlockIndex(hash)};
+    if (!index) return nullptr;
+
+    if (m_chainman.IsBlockInMainOrBestChain(hash) ||
+        m_chainman.IsBlockInChainTipSet(hash, GetChainTipSetForPeerNetwork(node))) {
```

Comment:
If we add the criteria that a block is observable by a peer if it is an ancestor of `pindexBestKnownBlock` for the peer that is giving it to us, then I think we can go back to keeping the `BlockRequestAllowed()` rule that I believe was originally in this PR, so that old blocks that are off the main chain are generally unobservable.

This should allow us to process a large reorg from a peer while providing better fingerprinting protection, by not allowing any blocks that are sufficiently old and off the main chain from being observable by anyone (other than peer that is in the middle of giving such a chain to us).  In particular this prevents fingerprinting a node over time -- how useful that is, if we are using the same network address anyway is admittedly not very clear to me(!), but it just seems like an easier heuristic to understand that we never allow old blocks that are off the main chain to be observable.

If we did that, then I'm not sure we need any kind of per-network map for evaluating observability; the rule could be that an entry in our block index is observable by a given peer if:
1) it is on the main chain, or
2) it has at last as much work as our tip, or
3) it forks off the main chain recently (ie the `BlockRequestAllowed` test), or
4) it is an ancestor of the peer's `pindexBestKnownBlock`

This seems to be sufficient for anti-fingerprinting while still allowing headers sync.

Reply:
I agree that using `BlockRequestAllowed` for this would be better as it comes with less complexity than what is in this PR. In practice, it would probably also always work out since we don't really expect large reorgs (1000s of blocks). However, I don't see how using `pindexBestKnownBlock` would solve the issues with that approach. `pindexBestKnownBlock` is of type `const CBlockIndex*`, so it can only point to something that already exists in our block index. We have to accept a header from a peer for us to set `pindexBestKnownBlock`.

The description in [this gist](https://gist.github.com/dergoegge/e8748c00ffe20b31108c95d88ba0ac73) should make it clear what the actual problem was. It is quite the edge case but I don't know how to get around it without keeping track of which chain tip a peer is allowed to build on. In the gist's example, `pindexBestKnownBlock` would still be set to block "z" for the peer that is sending us headers building on block "x", so the ancestor check for it does not seem to work here.

> In particular this prevents fingerprinting a node over time 

We could also reset the chain tip set once per month (or every n blocks) to achieve this.

Reply:
Ah right, if the peer's best known block has more work than the last header we receive from them when we're downloading headers towards their new tip, then we won't reset it until we get to a header that has more work.  So you're right, what I wrote doesn't work.

I think we could still get the job done with just a small amount of per-peer state.  If we stored the last locator we sent a peer in a getheaders message, then in theory we should permit observability of any header referenced in the locator by the peer that we sent the message to.  The downside to trying that now is that currently we don't limit the number of simultaneous in-flight getheaders messages, so it is slightly messy to keep track of (potentially) multiple locators.  But given that we currently generate our locators as being ancestors of some last header, perhaps it would be enough to essentially keep a set of chain tips per peer that represent the last header of some outstanding getheaders message, and then remove from that set after some timeout or after a getheaders response comes in that builds off such a tip?

Typically you'd expect the set to be empty or have one entry in it, but if somehow you triggered multiple outstanding getheaders messages, you'd be able to deal with it.  (I actually have a branch I'm working on that would try to prevent sending multiple getheaders simultaneously to a peer, perhaps we can discuss offline sometime.)

If the approach I'm suggesting is too complex for now, I also think it would be ok to proceed along the lines of what you have, which seems conceptually to be a strict improvement, and we could revisit this logic down the road.

Reply:
>  perhaps it would be enough to essentially keep a set of chain tips per peer that represent the last header of some outstanding getheaders message

I like this idea and maybe some of the current logic in this PR could be reused for that (mainly the chain tip set logic). But its not just headers that allowed for this fingerprinting, there are two more uses of `LookupBlockindex` that are problematic. i.e. when receiving invs for blocks and receiving compact blocks, see [here](https://github.com/bitcoin/bitcoin/blob/2cf8c2caea90d9a3c314ba5f88bb76b3b5273d62/src/net_processing.cpp#L3113-L3124) and [here](https://github.com/bitcoin/bitcoin/blob/2cf8c2caea90d9a3c314ba5f88bb76b3b5273d62/src/net_processing.cpp#L3601).

> If the approach I'm suggesting is too complex for now, I also think it would be ok to proceed along the lines of what you have, which seems conceptually to be a strict improvement, and we could revisit this logic down the road.

Yea, i would prefer keeping it as is and maybe revisit later on if we find an even better solution.

Reply:
> Yea, i would prefer keeping it as is and maybe revisit later on if we find an even better solution.

Agreed, the more I think about this the more I'm concluding that the improvements I'm thinking of would be much easier if we got to a place where we've improved the way we send getheaders messages (so that we don't have more than one in flight to a peer).  I think that would help with the situations where we're in the middle of headers sync and somehow receive an INV or a compact block, and have to reason about whether the header in question should be observable.  (The easiest answer, I think, is that if we're in the middle of headers sync we shouldn't be issuing a new getheaders in the first place!)

I guess the only downside is that I do think the per-network visibility rules introduced here may not be necessary in the future, but I think that's okay on the whole even if it turned out to be the case, as it would still be an improvement for now.  I'll try to focus the remainder of my review on this particular approach!  Thanks for the discussion.